src/librustc_middle/ty/relate.rs

```diff
@@ -559,7 +560,31 @@ pub fn super_relate_consts<R: TypeRelation<'tcx>>(
                     }
                 }
 
-                // FIXME(const_generics): handle `ConstValue::ByRef`.
+                (ConstValue::ByRef { .. }, ConstValue::ByRef { .. }) => {
+                    match a.ty.kind {
+                        ty::Array(..) | ty::Adt(..) | ty::Tuple(..) => {
+                            let a_destructured = tcx.destructure_const(relation.param_env().and(a));
+                            let b_destructured = tcx.destructure_const(relation.param_env().and(b));
+
+                            // Both the variant and each field have to be equal.
+                            if a_destructured.variant == b_destructured.variant {
+                                for (a_field, b_field) in
+                                    a_destructured.fields.iter().zip(b_destructured.fields.iter())
+                                {
+                                    relation.consts(a_field, b_field)?;
+                                }
+
+                                Ok(a_val)
+                            } else {
+                                Err(TypeError::ConstMismatch(expected_found(relation, &a, &b)))
+                            }
+                        }
+                        // FIXME(const_generics): There are probably some `TyKind`s
+                        // which should be handled here.
+                        _ => Err(TypeError::ConstMismatch(expected_found(relation, &a, &b))),
```

Comment:
I can't think of any, at least out of the types we allow. You should use at least `delay_span_bug`, because returning `Err` indicates that they're *definitely* not equal (which may allow e.g. overlapping `impl`s).

Reply:
:ok_hand: should `relate_consts` only be called for types which are structural match?

Is structural match implemented for `&`?

i.e. should this work
```rust
#![feature(const_generics)]

struct A<const N: &'static ()>;

fn main() {
    let _ = A::<{&()}>;
}
```

Reply:
Yes, it should, but it's blocked on one of @oli-obk's PRs.